### ISSUE TITLE
Separate Emacs, CUA and Vi maps

### ISSUE THREAD

**Issue by aartaka** [atlas-engineer/nyxt#1841]
As discussed in [this forum thread](https://discourse.atlas.engineer/t/c-w-while-in-a-textbox-makes-me-sad/315/7), having both CUA and Emacs bindings in Emacs mode (and VI, Emacs, and CUA in VI mode) causes confusion or even more serious consequences.

How about we separate CUA, Emacs, and VI maps?

**Comment by Ambrevar** [comment]
Inheritance is very convenient, but I understand why it may not suit everyone's needs.

What about importing instead?  For instance Emacs could import CUA bindings and _then_ unbind `C-w` if need be in the Emacs scheme.

The drawback is that it's less dynamic live, e.g. if something updates the CUA bindings while Nyxt is running, then change won't reflect on the Emacs bindings.  Not so much of problem in practice in my opinion.

Thoughts?

**Comment by jmercouris** [comment]
I think inheritance is a good default. How about instead of importing bindings, we let users redefine the inheritance hierarchy? For example, there is no need to make emacs inherit from CUA, it could just stand alone, no?

**Comment by Ambrevar** [comment]
The user can already to this by redefining the `emacs` scheme.  Currently the issue is that we take inheritance for granted and thus rely on some bindings being inherited from CUA (and Emacs for the VI scheme).

If we stick to inheritance, we could indeed follow @jmercouris' suggestion:
- Define all bindings explicitly for all schemes.
- Disable undesired bindings in Emacs and VI schemes (see https://github.com/atlas-engineer/nyxt/issues/1743).
- Document how to remove inheritance.

**Comment by Ambrevar** [comment]
A thought: the original issue seems to come from undesired bindings still being present when Emacs or VI is selected.
Is the problem simply that there are some bindings which should be in CUA but not in other schemes?

If so, here is a simple solution that would let us keep inheritance:  change the hierachy.

- default-scheme
  - CUA
  - Emacs
  - vi-insert
  - vi-normal (currently it depends on `emacs` but I think it shouldn't)

Actually I'm not too sure that vi-insert should depend on default-scheme either, so maybe it should be

- default-scheme
  - CUA
  - Emacs
  - vi-normal (currently it depends on `emacs` but I think it shouldn't)
- vi-insert

Thoughts?

**Comment by aartaka** [comment]
Indeed, `vi-insert-mode` seems to be distinct and mostly empty! Note that there's also `passthrough-mode`, which also influences the keys.

**Comment by Ambrevar** [comment]
Aren't you mixing up modes and keymap schemes?  I was only talking about keymap schemes here ;)

**Comment by aadcg** [comment]
Perhaps the issue here is that when the user asks to list all the keybindings he actually sees the inheritance hierarchy. If we change that, then it would be being a problem I guess.

About the proposed hierarchy, I agree @Ambrevar.

**Comment by Ambrevar** [comment]
@aadcg @aartaka @jmercouris Does someone have the time to address this?  :)

**Comment by aartaka** [comment]
Will do!

**Comment by aartaka** [comment]
But wait, what exactly do we have left to do? Schemes are now separated from each other in nkeymaps...

**Comment by Ambrevar** [comment]
Define `default` bindings! :)  For instance the `f1` help keys should go there.

**Comment by Ambrevar** [comment]
I had forgotten to add the `default` keyscheme.  It's now done with ac561d85025a955a3e95c059d9eed58b8190bcd3.

**Comment by jmercouris** [comment]
I'm having trouble seeing a large distinction between default and CUA. Is default simply something that exists in CUA but does not interfere with Emacs?

**Comment by Ambrevar** [comment]
Correct!  See the forum thread mentioned in the original post: `C-w` in VI mode used to close the buffer, but really it should not!

**Comment by jmercouris** [comment]
I guess this *must* be a manual process. Unless we could do some sort of set operations against the set of bindings... hm...

**Comment by Ambrevar** [comment]
What do you mean?
